C compilation error - cannot take the address of an rvalue of type 'u8'

<!-- It is advisable to update all relevant modules using `v outdated` and `v install` -->

**V doctor:**

```
V full version: V 0.4.8 97c6cd0
OS: macos, macOS, 12.7.6, 21H1320
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-4870HQ CPU @ 2.50GHz

getwd: /Users/dmitri/Dev/v_projects
vexe: /Users/dmitri/langs/v/v
vexe mtime: 2024-10-23 19:39:48

vroot: OK, value: /Users/dmitri/langs/v
VMODULES: OK, value: /Users/dmitri/.vmodules
VTMP: OK, value: /tmp/v_501

Git version: git version 2.37.1 (Apple Git-137.1)
Git vroot status: weekly.2024.22-767-g97c6cd09
.git/config present: true

CC version: Apple clang version 14.0.0 (clang-1400.0.29.202)
thirdparty/tcc status: thirdparty-macos-amd64 975f1ad8
```

**What did you do?**
`./v -g -o vdbg cmd/v && ./vdbg bug.v`

```v
fn main() {
	a := 'abcd'
	b := a[0].str
	println(typeof(b).name)
}
```

**What did you expect to see?**

type name of the 'b' variable

**What did you see instead?**

```
================== C compilation error (from cc): ==============
cc: /tmp/v_501/bug.01JAXEEYN87HG0A44RZBZPYSJM.tmp.c:13637:83: error: cannot take the address of an rvalue of type 'u8' (aka 'unsigned char')
cc:         string (*b) (void) = __closure_create(_V_closure_u8_str_36, memdup_uncollectable(&string_at(a, 0), sizeof(u8)));
cc:                                                                                          ^~~~~~~~~~~~~~~~
cc: 1 error generated.
================================================================
(You can pass `-cg`, or `-show-c-output` as well, to print all the C error messages).
builder error: 
==================
C error found. It should never happen, when compiling pure V code.
This is a V compiler bug, please report it using `v bug file.v`,
or goto https://github.com/vlang/v/issues/new/choose .
You can also use #help on Discord: https://discord.gg/vlang .
```

> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

<sub><a href="https://huly.app/guest/vlang-66f40c4d-a476b54c67-771fdd?token=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJsaW5rSWQiOiI2NzE5NTMxNGYwMmQ4YTAzZmIyZmU5NGEiLCJndWVzdCI6InRydWUiLCJlbWFpbCI6IiNndWVzdEBoYy5lbmdpbmVlcmluZyIsIndvcmtzcGFjZSI6InctYWxleGFuZGVyLXZsYW5nLTY2ZjQwYzRkLWE0NzZiNTRjNjctNzcxZmRjIn0.QEiMJLfmYZxxbt_p_xqSeNj2lKhQ7lGVBo7EGWhMf4w">Huly&reg;: <b>V_0.6-21084</b></a></sub>

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -4063,8 +4063,14 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 			if !receiver.typ.is_ptr() {
 				g.write('memdup_uncollectable(')
 			}
+			mut has_addr := false
 			if !node.expr_type.is_ptr() {
-				g.write('&')
+				if node.expr is ast.IndexExpr {
+					has_addr = true
+					g.write('ADDR(${g.styp(node.expr_type)}, ')
+				} else {
+					g.write('&')
+				}
 			}
 			if !node.expr.is_lvalue() {
 				current_stmt := g.go_before_last_stmt()
@@ -4074,6 +4080,9 @@ fn (mut g Gen) selector_expr(node ast.SelectorExpr) {
 				g.expr(ast.Expr(var))
 			} else {
 				g.expr(node.expr)
+			}
+			if has_addr {
+				g.write(')')
 			}
 			if !receiver.typ.is_ptr() {
 				g.write(', sizeof(${expr_styp}))')

--- a/vlib/v/tests/selector_indexexpr_fn_type_assign_test.v
+++ b/vlib/v/tests/selector_indexexpr_fn_type_assign_test.v
@@ -1,0 +1,11 @@
+fn test_main() {
+	a := 'abcd'
+	b := a[0].str
+	println(typeof(b).name)
+	assert b() == u8(`a`).str()
+
+	c := [1]
+	d := c[0].str
+	println(typeof(d).name)
+	assert d() == '1'
+}

